third_party/move/move-compiler-v2/src/pipeline/ability_checker.rs

```diff
@@ -0,0 +1,259 @@
+fn check_copy_for_temp_with_msg(
+    func_target: &FunctionTarget,
+    t: TempIndex,
+    loc: &Loc,
+    err_msg: &str,
+) {
+    let ty = func_target.get_local_type(t);
+    check_copy(func_target, ty, loc, err_msg)
+}
+
+fn check_read_ref(target: &FunctionTarget, t: TempIndex, loc: &Loc) {
+    if let Type::Reference(_, ty) = target.get_local_type(t) {
+        check_copy(target, ty, loc, "cannot copy")
+    } else {
+        panic!("ICE from ability checker: read_ref has non-reference argument")
+    }
+}
+
+/// Determines if the given type has the drop constraint.
+fn has_drop(func_target: &FunctionTarget, ty: &Type) -> bool {
+    type_abilities(func_target, ty).has_ability(Ability::Drop)
+}
+
+/// Checks if the given type has constraint drop, and add diagnostics if not.
+fn check_drop(func_target: &FunctionTarget, ty: &Type, loc: &Loc, err_msg: &str) {
+    if !has_drop(func_target, ty) {
+        func_target.global_env().error(loc, err_msg)
+    }
+}
+
+/// Checks if the given temporary variable has constraint drop, and add diagnostics if not.
+fn check_drop_for_temp_with_msg(
+    func_target: &FunctionTarget,
+    t: TempIndex,
+    loc: &Loc,
+    err_msg: &str,
+) {
+    let ty = func_target.get_local_type(t);
+    check_drop(func_target, ty, loc, err_msg)
+}
+
+/// `t` is the local containing the reference being written to
+fn check_write_ref(target: &FunctionTarget, t: TempIndex, loc: &Loc) {
+    if let Type::Reference(_, ty) = target.get_local_type(t) {
+        check_drop(target, ty, loc, "write_ref: cannot drop")
+    } else {
+        panic!("ICE typing error")
+    }
+}
+
+fn check_key_for_struct(
+    target: &FunctionTarget,
+    mod_id: ModuleId,
+    struct_id: StructId,
+    insts: &[Type],
+    loc: &Loc,
+    err_msg: &str,
+) {
+    if !check_struct_inst(target, mod_id, struct_id, insts, &loc).has_ability(Ability::Key) {
+        target.global_env().error(loc, err_msg)
+    }
+}
+
+fn ty_param_abilities(ty_params: &[TypeParameter]) -> impl Fn(u16) -> AbilitySet + Copy + '_ {
+    |i| {
+        if let Some(tp) = ty_params.get(i as usize) {
+            tp.1.abilities
+        } else {
+            panic!("ICE unbound type parameter")
+        }
+    }
+}
+
+fn get_abilities<'a>(
+    target: &'a FunctionTarget,
+) -> impl Fn(ModuleId, StructId) -> (Vec<AbilitySet>, AbilitySet) + Copy + 'a {
+    |mid, sid| {
+        let qid = QualifiedId {
+            module_id: mid,
+            id: sid,
+        };
+        let struct_env = target.global_env().get_struct(qid);
+        let struct_abilities = struct_env.get_abilities();
+        let params_ability_constraints = struct_env
+            .get_type_parameters()
+            .iter()
+            .map(|tp| tp.1.abilities)
+            .collect_vec();
+        (params_ability_constraints, struct_abilities)
+    }
+}
+
+/// checks that the given type is instantiated with types satisfying their ability constraints
+/// on the type parameter
```

Comment:
This comment doesn't quite make sense.  Is `sid` the "given type"?  Also, what is the return value?

Reply:
It's essentially a helper called by `check_inst(Type::Struct(mid, sid, ty_args)`.  Also mid, sid, ty_args are provided to generic instructions for instance `Pack(ModuleId, StructId, Vec<Type>)` as the type argument, not the struct type.

Reply:
done